### PR TITLE
cli: add probe command to show receiver register information

### PR DESCRIPTION
**Pull request by pfps**
I wrote this code to dump receiver registers. 

I don't know if it is a good idea to add it to Solaar.  I created this PR for discussion.

Addresses #74

### CODE REVIEW

**Comment by pfps** [comment]
Sample output:
> bin/solaar -D /dev/hidraw5 probe
Unifying Receiver
  Device path  : /dev/hidraw5
  USB id       : 046d:c52b
  Serial       : 1EB692A3
    Firmware   : 12.10.B0032
    Bootloader : 02.14
    Other      : AA.AA
  Has 2 paired device(s) out of a maximum of 6.
  Notifications: wireless, software present (0x000900)
  Device activity counters: 1=19

REGISTER DUMP
Notifications:       000900
Connections:         000200
Pairing State:       None
Device Activity:     13000000000000000000000000000000
Pairing Register  0: None
Pairing Register 10: None
Pairing Register 20: 20110810250400020C00000000000000
Pairing Register 30: 3093E8527F0400000001000000000000
Pairing Register 40: 40044D35313000000000000000000000
Pairing Name     40: b'M510'
Pairing Register  1: None
Pairing Register 11: None
Pairing Register 21: 21121420110400010D1A000000000000
Pairing Register 31: 31811119511A40000002000000000000
Pairing Register 41: 41044B35323020202020202020202020
Pairing Name     41: b'K520'
Pairing Register  2: 02121000328802040100000000000000
Pairing Register 12: None
Pairing Register 22: None
Pairing Register 32: None
Pairing Register 42: None
Pairing Name     42: None
Pairing Register  3: 031EB692A30106120000000000000000
Pairing Register 13: None
Pairing Register 23: None
Pairing Register 33: None
Pairing Register 43: None
Pairing Name     43: None
Pairing Register  4: None
Pairing Register 14: None
Pairing Register 24: None
Pairing Register 34: None
Pairing Register 44: None
Pairing Name     44: None
Pairing Register  5: None
Pairing Register 15: None
Pairing Register 25: None
Pairing Register 35: None
Pairing Register 45: None
Pairing Name     45: None

INTERPRETING SELECT REGISTER INFORMATION
Serial Number: 1EB692A3
Max Devices: 6
Can't write pairing state register so receiver can unpair (probably incorrect)
Number of paired devices: 2
Number of remaining pairings: Unlimited - raw value: 0x0

**Comment by doctor64** [comment]
Nice feature. May be it can be extended to read device registers (I mean HID++ 1.0 devices only), like old script `tools/scan-registers.sh`?

**Comment by pfps** [comment]
Extending this feature is certainly possible.   The problem is going to be what to probe.  The logitech information is incomplete even for the receiver stuff.  My receivers have more registers than probe probes but I don't have documentation on most of it.

File: lib/solaar/cli/probe.py [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+			raise Exception("no receiver found matching '%s'" % receiver_name)
+	else:
+		receiver = receivers[0]
+
+	assert receiver
+
+	_print_receiver(receiver)
+
+	print("")
+	print("REGISTER DUMP")
```
Comment (FFY00) [inline lib/solaar/cli/probe.py]: No caps here (`Register Dump`) and match the level (`' ' * 2`).

File: lib/solaar/cli/probe.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+
+	register = receiver.read_register(_R.notifications)
+	print("Notification Register: %s" % (_strhex(register) if register else "None"))
+	register = receiver.read_register(_R.receiver_connection)
+	print("Connection State:      %s" % (_strhex(register) if register else "None"))
+	register = receiver.read_register(_R.receiver_pairing)
+	print("Pairing State:         %s" % (_strhex(register) if register else "None"))
+	register = receiver.read_register(_R.devices_activity)
+	print("Device Activity:       %s" % (_strhex(register) if register else "None"))
```
Comment (FFY00) [inline lib/solaar/cli/probe.py]: One more level (`' ' * 4`). Same for the reset of the code.

File: lib/solaar/cli/probe.py [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+		if not receiver:
+			raise Exception("no receiver found matching '%s'" % receiver_name)
+	else:
+		receiver = receivers[0]
+
+	assert receiver
+
+	_print_receiver(receiver)
+
+	print("")
```
Comment (FFY00) [inline lib/solaar/cli/probe.py]: Remove please.

File: lib/solaar/cli/probe.py [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+	print("REGISTER DUMP")
+
+	register = receiver.read_register(_R.notifications)
+	print("Notification Register: %s" % (_strhex(register) if register else "None"))
```
Comment (FFY00) [inline lib/solaar/cli/probe.py]: `0x` prefix on hex numbers.

File: lib/solaar/cli/probe.py (hunk 1 above)
Comment (pfps) [reply lib/solaar/cli/probe.py]: Done

File: lib/solaar/cli/probe.py (hunk 2 above)
Comment (pfps) [reply lib/solaar/cli/probe.py]: Done

File: lib/solaar/cli/probe.py (hunk 3 above)
Comment (pfps) [reply lib/solaar/cli/probe.py]: Removed empty line

File: lib/solaar/cli/probe.py (hunk 4 above)
Comment (pfps) [reply lib/solaar/cli/probe.py]: Done, and added hex value for registers.

**Comment by pfps** [comment]
I also remove the register interpretation section.  It wasn't adding anything significant.

**Comment by pfps** [comment]
New output format:

```
Unifying Receiver
  Device path  : /dev/hidraw5
  USB id       : 046d:c52b
  Serial       : 1EB692A3
    Firmware   : 12.10.B0032
    Bootloader : 02.14
    Other      : AA.AA
  Has 2 paired device(s) out of a maximum of 6.
  Notifications: wireless, software present (0x000900)
  Device activity counters: 1=191
  Register Dump
    Notification Register 0x00: 0x000900
    Connection State      0x02: 0x000200
    Device Activity       0xb3: 0xBF000000000000000000000000000000
    Pairing Register 0xb5 0x00: None
    Pairing Register 0xb5 0x10: None
    Pairing Register 0xb5 0x20: 0x20110810250400020C00000000000000
    Pairing Register 0xb5 0x30: 0x3093E8527F0400000001000000000000
    Pairing Name     0xb5 0x40: b'M510'
    Pairing Register 0xb5 0x01: None
    Pairing Register 0xb5 0x11: None
    Pairing Register 0xb5 0x21: 0x21121420110400010D1A000000000000
    Pairing Register 0xb5 0x31: 0x31811119511A40000002000000000000
    Pairing Name     0xb5 0x41: b'K520'
    Pairing Register 0xb5 0x02: 0x02121000328802040100000000000000
    Pairing Register 0xb5 0x12: None
    Pairing Register 0xb5 0x22: None
    Pairing Register 0xb5 0x32: None
    Pairing Name     0xb5 0x42: None
    Pairing Register 0xb5 0x03: 0x031EB692A30106120000000000000000
    Pairing Register 0xb5 0x13: None
    Pairing Register 0xb5 0x23: None
    Pairing Register 0xb5 0x33: None
    Pairing Name     0xb5 0x43: None
    Pairing Register 0xb5 0x04: None
    Pairing Register 0xb5 0x14: None
    Pairing Register 0xb5 0x24: None
    Pairing Register 0xb5 0x34: None
    Pairing Name     0xb5 0x44: None
    Pairing Register 0xb5 0x05: None
    Pairing Register 0xb5 0x15: None
    Pairing Register 0xb5 0x25: None
    Pairing Register 0xb5 0x35: None
    Pairing Name     0xb5 0x45: None
    Firmware         0xf1 0x00: None
    Firmware         0xf1 0x01: 0x011210
    Firmware         0xf1 0x02: 0x020032
    Firmware         0xf1 0x03: 0x03AAAA
    Firmware         0xf1 0x04: 0x040214
```

**Pull request merged**